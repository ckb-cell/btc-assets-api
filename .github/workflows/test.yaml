---
name: Unit Tests

on:
  pull_request:
    branches:
      - main
      - develop

jobs:
  test:
    runs-on: ubuntu-latest

    services:
      redis:
        image: redis
        options: >-
          --health-cmd "redis-cli ping"
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5
        ports:
          - 6379:6379

    steps:
      - name: Checkout Code
        uses: actions/checkout@v3

      - name: Install PNPM
        uses: pnpm/action-setup@v3
        with:
          version: 8

      - name: Use Node.js
        uses: actions/setup-node@v3
        with:
          node-version: "21"
          cache: "pnpm"

      - name: Install dependencies
        run: pnpm install

      - name: Run Unit Tests
        env:
          JWT_SECRET: ${{ secrets.JWT_SECRET }}
          BITCOIN_JSON_RPC_URL: ${{ secrets.BITCOIN_JSON_RPC_URL }}
          BITCOIN_JSON_RPC_USERNAME: ${{ secrets.BITCOIN_JSON_RPC_USERNAME }}
          BITCOIN_JSON_RPC_PASSWORD: ${{ secrets.BITCOIN_JSON_RPC_PASSWORD }}
          BITCOIN_ELECTRS_API_URL: ${{ secrets.BITCOIN_ELECTRS_API_URL }}
          BITCOIN_SPV_SERVICE_URL: ${{ secrets.BITCOIN_SPV_SERVICE_URL }}
          CKB_RPC_URL: ${{ secrets.CKB_RPC_URL }}
          CKB_INDEXER_URL: ${{ secrets.CKB_INDEXER_URL }}
          PAYMASTER_PRIVATE_KEY: ${{ secrets.PAYMASTER_PRIVATE_KEY }}
          REDIS_URL: redis://localhost:6379
          CI_REDIS_URL: redis://localhost:6379
        run: pnpm test
