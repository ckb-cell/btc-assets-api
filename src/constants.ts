export const CUSTOM_HEADERS = {
  ApiCache: 'x-api-cache',
  ResponseCacheable: 'x-response-cacheable',
};

export enum ApiCacheStatus {
  Hit = 'HIT',
  Miss = 'MISS',
}

export const JWT_IGNORE_URLS = ['/token', '/docs'];
