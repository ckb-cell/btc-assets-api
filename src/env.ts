import 'dotenv/config';
import z from 'zod';
import process from 'node:process';
import { omit } from 'lodash';

const envSchema = z.object({
  NODE_ENV: z.string().default('development'),
  PORT: z.string().optional(),
  ADDRESS: z.string().optional(),
  NETWORK: z.enum(['mainnet', 'testnet']).default('testnet'),

  DOMAIN: z.string().optional(),

  SENTRY_DSN_URL: z.string().optional(),
  REDIS_URL: z.string(),
  RATE_LIMIT_PER_MINUTE: z.number().default(100),
  LOGGER_LEVEL: z.enum(['debug', 'info', 'warn', 'error']).default('info'),

  ADMIN_USERNAME: z.string().optional(),
  ADMIN_PASSWORD: z.string().optional(),

  /**
   * JWT_SECRET is used to sign the JWT token for authentication.
   */
  JWT_SECRET: z.string(),
  /**
   * The URL/USERNAME/PASSWORD of the Bitcoin JSON-RPC server.
   * The JSON-RPC server is used to query the Bitcoin blockchain.
   */
  BITCOIN_JSON_RPC_URL: z.string(),
  BITCOIN_JSON_RPC_USERNAME: z.string(),
  BITCOIN_JSON_RPC_PASSWORD: z.string(),
  /**
   * The URL of the Electrs API.
   * Electrs is a Rust implementation of Electrum Server.
   * It is used to query the Bitcoin blockchain (balance, transactions, etc).
   */
  BITCOIN_ELECTRS_API_URL: z.string(),

  /**
   * Bitcoin SPV service URL
   * https://github.com/ckb-cell/ckb-bitcoin-spv-service
   */
  BITCOIN_SPV_SERVICE_URL: z.string(),
  /**
   * The URL of the CKB JSON-RPC server.
   */
  CKB_RPC_URL: z.string(),
  /**
   * Paymaster private key, used to sign the transaction with paymaster cell.
   */
  PAYMASTER_PRIVATE_KEY: z.string(),
  /**
   * Paymaster cell capacity in shannons
  * (254 CKB for RGB++ capacity + 61 CKB for change cell capacity + 1 CKB for fee cell)
   */
  PAYMASTER_CELL_CAPACITY: z.coerce.number().default(316 * 10 ** 8),
  PAYMASTER_CELL_PRESET_COUNT: z.coerce.number().default(500),
  PAYMASTER_CELL_REFILL_THRESHOLD: z.coerce.number().default(0.3),

  UNLOCKER_CRON_SCHEDULE: z.string().default('*/5 * * * *'),
  UNLOCKER_CELL_BATCH_SIZE: z.coerce.number().default(100),
  UNLOCKER_MONITOR_SLUG: z.string().default('btctimelock-cells-unlock'),

  TRANSACTION_QUEUE_JOB_DELAY: z.coerce.number().default(120 * 1000),
});

export type Env = z.infer<typeof envSchema>;
export const env = envSchema.parse(process.env);

export const getSafeEnvs = () =>
  omit(env, ['ADMIN_PASSWORD', 'JWT_SECRET', 'BITCOIN_JSON_RPC_PASSWORD', 'PAYMASTER_PRIVATE_KEY']);

export const isGenerateTokenPrivate = env.NODE_ENV === 'production' && env.ADMIN_USERNAME && env.ADMIN_PASSWORD;
