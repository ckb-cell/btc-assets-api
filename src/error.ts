export enum AppErrorCode {
  UnknownResponseError = 0x9999,
}
