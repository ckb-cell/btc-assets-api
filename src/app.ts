import { FastifyInstance } from 'fastify';
import { AxiosError } from 'axios';
import bitcoinRoutes from './routes/bitcoin';
import { env } from './env';
import tokenRoutes from './routes/token';
import swaggerPlugin from './plugins/swagger';
import jwtPlugin from './plugins/jwt';
import cachePlugin from './plugins/cache';
import rateLimitPlugin from './plugins/rate-limit';
import * as Sentry from '@sentry/node';

if (env.SENTRY_DSN_URL && env.NODE_ENV !== 'development') {
  Sentry.init({
    dsn: env.SENTRY_DSN_URL,
    tracesSampleRate: 1.0,
  });
}

async function routes(fastify: FastifyInstance) {
  fastify.register(rateLimitPlugin);
  fastify.register(swaggerPlugin);
  fastify.register(jwtPlugin);
  fastify.register(cachePlugin);

  fastify.register(tokenRoutes, { prefix: '/token' });
  fastify.register(bitcoinRoutes, { prefix: '/bitcoin/v1' });

  fastify.setErrorHandler((error, _, reply) => {
    Sentry.captureException(error);
    if (error instanceof AxiosError) {
      reply
        .status(error.response?.status || 500)
        .send({ ok: false, error: error.response?.data.error ?? error.message });
      return;
    }
    reply
      .status(error.statusCode ?? 500)
      .send({ ok: false, statusCode: error.statusCode, message: error.message });
  });
}

export default routes;
