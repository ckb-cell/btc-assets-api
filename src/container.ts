import { createContainer, InjectionMode, asValue, asClass, asFunction } from 'awilix';
import { Redis } from 'ioredis';
import pino from 'pino';
import Bitcoind from './services/bitcoind';
import ElectrsAPI from './services/electrs';
import { env } from './env';
import TransactionManager from './services/transaction';
import Paymaster from './services/paymaster';
import { RPC as CkbRPC, Indexer as CkbIndexer } from '@ckb-lumos/lumos';
import Unlocker from './services/unlocker';

export interface Cradle {
  env: typeof env;
  logger: pino.BaseLogger;
  redis: Redis;
  ckbRpc: CkbRPC;
  ckbIndexer: CkbIndexer;
  bitcoind: Bitcoind;
  electrs: ElectrsAPI;
  paymaster: Paymaster;
  unlocker: Unlocker;
  transactionManager: TransactionManager;
}

const container = createContainer<Cradle>({
  injectionMode: InjectionMode.PROXY,
  strict: true,
});

container.register({
  env: asValue(env),
  logger: asValue(pino()),
  redis: asValue(
    new Redis(env.REDIS_URL, {
      maxRetriesPerRequest: null,
    }),
  ),
  ckbRpc: asFunction(() => new CkbRPC(env.CKB_RPC_URL)).singleton(),
  ckbIndexer: asFunction(() => new CkbIndexer(env.CKB_INDEXER_URL)).singleton(),
  bitcoind: asClass(Bitcoind).singleton(),
  electrs: asClass(ElectrsAPI).singleton(),
  paymaster: asClass(Paymaster).singleton(),
  transactionManager: asClass(TransactionManager).singleton(),
  unlocker: asClass(Unlocker).singleton(),
});

export default container;
