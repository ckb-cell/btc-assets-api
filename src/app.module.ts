import { Module } from '@nestjs/common';
import { AppService } from './app.service';
import { TokenModule } from './modules/token/token.module';
import { BitcoinModule } from './modules/bitcoin/bitcoin.module';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { envSchema } from './env';
import { SentryModule } from '@ntegral/nestjs-sentry';
import { nodeProfilingIntegration } from '@sentry/profiling-node';
import { CacheModule } from '@nestjs/cache-manager';
import type { RedisClientOptions } from 'redis';
import * as redisStore from 'cache-manager-redis-store';
import { StoreConfig } from 'cache-manager';

@Module({
  imports: [
    ConfigModule.forRoot({
      isGlobal: true,
      validate: envSchema.parse,
    }),
    CacheModule.registerAsync<RedisClientOptions>({
      isGlobal: true,
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: async (configService: ConfigService) =>
        ({
          store: redisStore,
          url: configService.get('REDIS_URL'),
          ttl: 10,
        }) as StoreConfig,
    }),
    SentryModule.forRootAsync({
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: async (configService: ConfigService) => ({
        dsn: configService.get('SENTRY_DSN'),
        environment: configService.get('NODE_ENV'),
        tracesSampleRate: 0.5,
        profilesSampleRate: 0.5,
        integrations: [nodeProfilingIntegration()],
        logLevels:
          configService.get('NODE_ENV') === 'production'
            ? ['warn', 'error']
            : ['debug', 'log', 'warn', 'error'],
        beforeSend: (event) => {
          if (event.level === 'error') {
            return event;
          }
          return null;
        },
      }),
    }),
    TokenModule,
    BitcoinModule,
  ],
  controllers: [],
  providers: [AppService],
})
export class AppModule {}
