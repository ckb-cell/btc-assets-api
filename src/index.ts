import process from 'node:process';
import { env } from './env';
import { buildFastify } from './app';

const port = parseInt(env.PORT || '3000', 10);
const host = env.ADDRESS || '0.0.0.0';

const app = buildFastify();

app.listen({ port, host }, (err, address) => {
  if (err) {
    console.error(err);
    process.exit(1);
  }

  // eslint-disable-next-line no-console
  console.log(`Server listening at ${address}`);
});
