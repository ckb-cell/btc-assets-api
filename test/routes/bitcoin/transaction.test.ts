import { beforeEach, expect, test } from 'vitest';
import { buildFastify } from '../../../src/app';
import { describe } from 'node:test';

let token: string;

describe('/bitcoin/v1/transaction', () => {
  beforeEach(async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const response = await fastify.inject({
      method: 'POST',
      url: '/token/generate',
      payload: {
        app: 'test',
        domain: 'test.com',
      },
    });
    const data = response.json();
    token = data.token;

    await fastify.close();
  });

  test('Get transaction by txid', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const response = await fastify.inject({
      method: 'GET',
      url: '/bitcoin/v1/transaction/9706131c1e327a068a6aafc16dc69a46c50bc7c65f180513896bdad39a6babfc',
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = response.json();

    expect(response.statusCode).toBe(200);
    expect(data).toMatchSnapshot();

    await fastify.close();
  });

  test('Get not exists transaction', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const response = await fastify.inject({
      method: 'GET',
      url: '/bitcoin/v1/transaction/9706131c1e327a068a6aafc16dc69a46c50bc7c65f180513896bdad39a6babf1',
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = await response.json();

    expect(response.statusCode).toBe(404);
    expect(data).toMatchSnapshot();

    await fastify.close();
  });

  test('Get transaction hex', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const response = await fastify.inject({
      method: 'GET',
      url: '/bitcoin/v1/transaction/9706131c1e327a068a6aafc16dc69a46c50bc7c65f180513896bdad39a6babfc/hex',
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
    });
    const data = await response.json();

    expect(response.statusCode).toBe(200);
    expect(data).toEqual({
      hex: '020000000001013115e849359a46737e7bf3e121a5a5de666a1ad01933e3df7dc48f97199f18d00000000000ffffffff02e803000000000000160014dd72437bde53e22df65feb845a1dd35784f3e66cacc1020000000000160014f8d0addc86183d385061ee80c1b16b1975eacf4202483045022100f50e2e6ee5ea04e4f7029d29a28e80266969f91357055e1d53b0715b768c448102204ccd9dbb38d0cc007e07b9c8b4088d4fac4ab65fef49774bafe27acbfc3d1111012102d05848540f152d730e272bc9628dd4e4a5f4126fdf118ad6a4fd6afd26b0831300000000',
    });

    await fastify.close();
  });

  test('Send exists raw transaction', async () => {
    const fastify = buildFastify();
    await fastify.ready();

    const response = await fastify.inject({
      method: 'POST',
      url: '/bitcoin/v1/transaction',
      headers: {
        Authorization: `Bearer ${token}`,
        Origin: 'https://test.com',
      },
      body: {
        txhex:
          '02000000000101fe7b9cd0f75741e2ec1e3a6142eab945e64fab0ef15de4a66c635c0a789e986f0100000000ffffffff02e803000000000000160014dbf4360c0791098b0b14679e5e78015df3f2caad6a88000000000000160014dbf4360c0791098b0b14679e5e78015df3f2caad02473044022065829878f51581488f44c37064b46f552ea7354196fae5536906797b76b370bf02201c459081578dc4e1098fbe3ab68d7d56a99e8e9810bf2806d10053d6b36ffa4d0121037dff8ff2e0bd222690d785f9277e0c4800fc88b0fad522f1442f21a8226253ce00000000',
      },
    });

    expect(response.statusCode).toBe(400);
    await fastify.close();
  });
});
