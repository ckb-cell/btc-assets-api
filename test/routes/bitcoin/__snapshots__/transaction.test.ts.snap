// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Get transaction by txid 1`] = `
{
  "fee": 141,
  "locktime": 0,
  "size": 223,
  "status": {
    "block_hash": "000000000000000cef4a1d6264fe63f543128518a466d31c7e2a8d6395b52522",
    "block_height": 2579043,
    "block_time": 1708580004,
    "confirmed": true,
  },
  "txid": "9706131c1e327a068a6aafc16dc69a46c50bc7c65f180513896bdad39a6babfc",
  "version": 2,
  "vin": [
    {
      "is_coinbase": false,
      "prevout": {
        "scriptpubkey": "0014f8d0addc86183d385061ee80c1b16b1975eacf42",
        "scriptpubkey_address": "tb1qlrg2mhyxrq7ns5rpa6qvrvttr9674n6z0trymp",
        "scriptpubkey_asm": "OP_0 OP_PUSHBYTES_20 f8d0addc86183d385061ee80c1b16b1975eacf42",
        "scriptpubkey_type": "v0_p2wpkh",
        "value": 181793,
      },
      "scriptsig": "",
      "scriptsig_asm": "",
      "sequence": 4294967295,
      "txid": "d0189f19978fc47ddfe33319d01a6a66dea5a521e1f37b7e73469a3549e81531",
      "vout": 0,
      "witness": [
        "3045022100f50e2e6ee5ea04e4f7029d29a28e80266969f91357055e1d53b0715b768c448102204ccd9dbb38d0cc007e07b9c8b4088d4fac4ab65fef49774bafe27acbfc3d111101",
        "02d05848540f152d730e272bc9628dd4e4a5f4126fdf118ad6a4fd6afd26b08313",
      ],
    },
  ],
  "vout": [
    {
      "scriptpubkey": "0014dd72437bde53e22df65feb845a1dd35784f3e66c",
      "scriptpubkey_address": "tb1qm4eyx777203zmajlawz958wn27z08envm2jelm",
      "scriptpubkey_asm": "OP_0 OP_PUSHBYTES_20 dd72437bde53e22df65feb845a1dd35784f3e66c",
      "scriptpubkey_type": "v0_p2wpkh",
      "value": 1000,
    },
    {
      "scriptpubkey": "0014f8d0addc86183d385061ee80c1b16b1975eacf42",
      "scriptpubkey_address": "tb1qlrg2mhyxrq7ns5rpa6qvrvttr9674n6z0trymp",
      "scriptpubkey_asm": "OP_0 OP_PUSHBYTES_20 f8d0addc86183d385061ee80c1b16b1975eacf42",
      "scriptpubkey_type": "v0_p2wpkh",
      "value": 180652,
    },
  ],
  "weight": 562,
}
`;
